src/rpc/misc.cpp

```diff
@@ -320,6 +320,86 @@ UniValue createmultisig(const JSONRPCRequest& request)
+{
+    if (request.fHelp || request.params.size() < 5 || request.params.size() > 6)
+        throw std::runtime_error(
+            "getsignaturehash \"tx\" \"scriptcode\" \"amount\" \"sigversion\" inputindex (\"sighashtype\")\n"
+            "\nCreate the hash to sign for the given input\n"
+
+            "\nArguments:\n"
+            "1. \"tx\"     (string, required) The transaction hex string\n"
+            "2. \"scriptcode\"     (string, required) The scriptCode to sign\n"
+            "3. \"amount\"     (numeric, required) The amount spent\n"
+            "4. \"sigversion\"     (string, required) The signature version\n"
+            "       \"BASE\"\n"
+            "       \"WITNESS_V0\"\n"
+            "5. inputindex                  (numeric, required) The index of the input to sign\n"
+            "6. \"sighashtype\"     (string, optional, default=ALL) The signature hash type. Must be one of\n"
+            "       \"ALL\"\n"
+            "       \"NONE\"\n"
+            "       \"SINGLE\"\n"
+            "       \"ALL|ANYONECANPAY\"\n"
+            "       \"NONE|ANYONECANPAY\"\n"
+            "       \"SINGLE|ANYONECANPAY\"\n"
+            "\nResult:\n"
+            "\"signature\"              (string) hex string of the signature\n"
+
+            "\nExamples:\n"
+            + HelpExampleCli("getsignaturehash", "\"txhex\" \"scriptcode\" \"0.01\" \"WITNESS_V0\" 1 \"SINGLE\"")
+            + HelpExampleRpc("getsignaturehash", "\"txhex\" \"scriptcode\" \"0.01\" \"WITNESS_V0\" 1 \"SINGLE\"")
+        );
+
+    RPCTypeCheck(request.params, {
+        UniValue::VSTR, // tx
+        UniValue::VSTR, // scriptCode
+        UniValue::VSTR, // amount
+        UniValue::VSTR, // sigVersion
+        UniValue::VNUM // inputIndex
+    },
+        false);
+
+    CMutableTransaction mtx;
+    if (!DecodeHexTx(mtx, request.params[0].get_str(), true))
+        throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "TX decode failed");
+
+    std::vector<unsigned char> scriptCodeData(ParseHexV(request.params[1], "scriptCode"));
+    CScript scriptCode(scriptCodeData.begin(), scriptCodeData.end());
+
+    auto amount = AmountFromValue(request.params[2].get_str());
+    SigVersion sigVersion;
+    static std::map<std::string, SigVersion> mapSigVersionValues = {
+        { std::string("BASE"), SigVersion::SIGVERSION_BASE },
+        { std::string("WITNESS_V0"), SigVersion::SIGVERSION_WITNESS_V0 },
+    };
+    std::string strHashType = request.params[3].get_str();
+    if (mapSigVersionValues.count(strHashType))
+        sigVersion = mapSigVersionValues[strHashType];
+    else
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid sigversion param");
+
+    int inputIndex = request.params[4].get_int();
+    int nHashType = SIGHASH_ALL;
+    if (!request.params[5].isNull()) {
+        RPCTypeCheckArgument(request.params[5], UniValue::VSTR);
+        static std::map<std::string, int> mapSigHashValues = {
+            { std::string("ALL"), int(SIGHASH_ALL) },
```

Comment:
This table is duplicated with signrawtransaction, which makes it harder to change later. Please factor it out.